// nx/src/devkit-internals.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.readModulePackageJson = exports.stripIndent = exports.sortObjectByKeys = exports.combineOptionsForExecutor = exports.splitTarget = exports.retrieveProjectConfigurationsWithAngularProjects = exports.calculateDefaultProjectName = exports.readNxJsonFromDisk = exports.getExecutorInformation = exports.createTempNpmDirectory = void 0;
/**
 * Note to developers: STOP! These exports are available via requireNx in @nx/devkit.
 *
 * These may not be available in certain version of Nx, so be sure to check them first.
 */
var package_manager_1 = require("./utils/package-manager");
Object.defineProperty(exports, "createTempNpmDirectory", { enumerable: true, get: function () { return package_manager_1.createTempNpmDirectory; } });
var executor_utils_1 = require("./command-line/run/executor-utils");
Object.defineProperty(exports, "getExecutorInformation", { enumerable: true, get: function () { return executor_utils_1.getExecutorInformation; } });
var nx_json_1 = require("./config/nx-json");
Object.defineProperty(exports, "readNxJsonFromDisk", { enumerable: true, get: function () { return nx_json_1.readNxJson; } });
var calculate_default_project_name_1 = require("./config/calculate-default-project-name");
Object.defineProperty(exports, "calculateDefaultProjectName", { enumerable: true, get: function () { return calculate_default_project_name_1.calculateDefaultProjectName; } });
var retrieve_workspace_files_1 = require("./project-graph/utils/retrieve-workspace-files");
Object.defineProperty(exports, "retrieveProjectConfigurationsWithAngularProjects", { enumerable: true, get: function () { return retrieve_workspace_files_1.retrieveProjectConfigurationsWithAngularProjects; } });
var split_target_1 = require("./utils/split-target");
Object.defineProperty(exports, "splitTarget", { enumerable: true, get: function () { return split_target_1.splitTarget; } });
var params_1 = require("./utils/params");
Object.defineProperty(exports, "combineOptionsForExecutor", { enumerable: true, get: function () { return params_1.combineOptionsForExecutor; } });
var object_sort_1 = require("./utils/object-sort");
Object.defineProperty(exports, "sortObjectByKeys", { enumerable: true, get: function () { return object_sort_1.sortObjectByKeys; } });
var logger_1 = require("./utils/logger");
Object.defineProperty(exports, "stripIndent", { enumerable: true, get: function () { return logger_1.stripIndent; } });
var package_json_1 = require("./utils/package-json");
Object.defineProperty(exports, "readModulePackageJson", { enumerable: true, get: function () { return package_json_1.readModulePackageJson; } });
