// nx/src/command-line/init/init-v2.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.initHandler = void 0;
const fs_1 = require("fs");
const semver_1 = require("semver");
const output_1 = require("../../utils/output");
const package_manager_1 = require("../../utils/package-manager");
const add_nx_scripts_1 = require("./implementation/dot-nx/add-nx-scripts");
const child_process_1 = require("../../utils/child-process");
const fileutils_1 = require("../../utils/fileutils");
const versions_1 = require("../../utils/versions");
const utils_1 = require("./implementation/utils");
const enquirer_1 = require("enquirer");
const child_process_2 = require("child_process");
const angular_1 = require("./implementation/angular");
const workspace_context_1 = require("../../utils/workspace-context");
const connect_to_nx_cloud_1 = require("../connect/connect-to-nx-cloud");
const add_nx_to_npm_repo_1 = require("./implementation/add-nx-to-npm-repo");
const add_nx_to_monorepo_1 = require("./implementation/add-nx-to-monorepo");
async function initHandler(options) {
    const version = process.env.NX_VERSION ?? ((0, semver_1.prerelease)(versions_1.nxVersion) ? 'next' : 'latest');
    if (process.env.NX_VERSION) {
        output_1.output.log({ title: `Using version ${process.env.NX_VERSION}` });
    }
    if (!(0, fs_1.existsSync)('package.json') || options.useDotNxInstallation) {
        if (process.platform !== 'win32') {
            console.log('Setting Nx up installation in `.nx`. You can run Nx commands like: `./nx --help`');
        }
        else {
            console.log('Setting Nx up installation in `.nx`. You can run Nx commands like: `./nx.bat --help`');
        }
        (0, add_nx_scripts_1.generateDotNxSetup)(version);
        // invokes the wrapper, thus invoking the initial installation process
        (0, child_process_1.runNxSync)('');
        return;
    }
    // TODO(jack): Remove this Angular logic once `@nx/angular` is compatible with inferred targets.
    if ((0, fs_1.existsSync)('angular.json')) {
        await (0, angular_1.addNxToAngularCliRepo)({
            ...options,
            integrated: !!options.integrated,
        });
        return;
    }
    const detectPluginsResponse = await detectPlugins();
    if (!detectPluginsResponse?.plugins.length) {
        // If no plugins are detected/chosen, guide users to setup
        // their targetDefaults correctly so their package scripts will work.
        const packageJson = (0, fileutils_1.readJsonFile)('package.json');
        if ((0, utils_1.isMonorepo)(packageJson)) {
            await (0, add_nx_to_monorepo_1.addNxToMonorepo)({ interactive: options.interactive });
        }
        else {
            await (0, add_nx_to_npm_repo_1.addNxToNpmRepo)({ interactive: options.interactive });
        }
    }
    else {
        const useNxCloud = options.nxCloud ??
            (options.interactive
                ? await (0, connect_to_nx_cloud_1.connectExistingRepoToNxCloudPrompt)()
                : false);
        const repoRoot = process.cwd();
        const pmc = (0, package_manager_1.getPackageManagerCommand)();
        (0, utils_1.createNxJsonFile)(repoRoot, [], [], {});
        (0, utils_1.updateGitIgnore)(repoRoot);
        (0, utils_1.addDepsToPackageJson)(repoRoot, detectPluginsResponse?.plugins ?? []);
        output_1.output.log({ title: 'ðŸ“¦ Installing Nx' });
        (0, utils_1.runInstall)(repoRoot, pmc);
        if (detectPluginsResponse) {
            output_1.output.log({ title: 'ðŸ”¨ Configuring plugins' });
            for (const plugin of detectPluginsResponse.plugins) {
                (0, child_process_2.execSync)(`${pmc.exec} nx g ${plugin}:init --keepExistingVersions ${detectPluginsResponse.updatePackageScripts
                    ? '--updatePackageScripts'
                    : ''} --no-interactive`, {
                    stdio: [0, 1, 2],
                    cwd: repoRoot,
                });
            }
        }
        if (useNxCloud) {
            output_1.output.log({ title: 'ðŸ› ï¸ Setting up Nx Cloud' });
            (0, child_process_2.execSync)(`${pmc.exec} nx g nx:connect-to-nx-cloud --installationSource=nx-init --quiet --hideFormatLogs --no-interactive`, {
                stdio: [0, 1, 2],
                cwd: repoRoot,
            });
        }
    }
    output_1.output.log({
        title: 'ðŸ‘€ Explore Your Workspace',
        bodyLines: [
            `Run "nx graph" to show the graph of the workspace. It will show tasks that you can run with Nx.`,
            `Read this guide on exploring your workspace: https://nx.dev/core-features/explore-graph`,
        ],
    });
}
exports.initHandler = initHandler;
const npmPackageToPluginMap = {
    // Generic JS tools
    eslint: '@nx/eslint',
    storybook: '@nx/storybook',
    // Bundlers
    vite: '@nx/vite',
    vitest: '@nx/vite',
    webpack: '@nx/webpack',
    // Testing tools
    jest: '@nx/jest',
    cypress: '@nx/cypress',
    '@playwright/test': '@nx/playwright',
    // Frameworks
    detox: '@nx/detox',
    expo: '@nx/expo',
    next: '@nx/next',
    nuxt: '@nx/nuxt',
    'react-native': '@nx/react-native',
    '@remix-run/dev': '@nx/remix',
};
async function detectPlugins() {
    let files = ['package.json'].concat((0, workspace_context_1.globWithWorkspaceContext)(process.cwd(), ['**/*/package.json']));
    const detectedPlugins = new Set();
    for (const file of files) {
        if (!(0, fs_1.existsSync)(file))
            continue;
        let packageJson;
        try {
            packageJson = (0, fileutils_1.readJsonFile)(file);
        }
        catch {
            // Could have malformed JSON for unit tests, etc.
            continue;
        }
        const deps = {
            ...packageJson.dependencies,
            ...packageJson.devDependencies,
        };
        for (const [dep, plugin] of Object.entries(npmPackageToPluginMap)) {
            if (deps[dep]) {
                detectedPlugins.add(plugin);
            }
        }
    }
    const plugins = Array.from(detectedPlugins);
    if (plugins.length === 0)
        return undefined;
    output_1.output.log({
        title: `Recommended Plugins:`,
        bodyLines: [
            `Add these Nx plugins to integrate with the tools used in your workspace.`,
        ],
    });
    const pluginsToInstall = await (0, enquirer_1.prompt)([
        {
            name: 'plugins',
            type: 'multiselect',
            message: `Which plugins would you like to add?`,
            choices: plugins.map((p) => ({ name: p, value: p })),
            initial: plugins.map((_, i) => i), // casting to avoid type error due to bad d.ts file from enquirer
        },
    ]).then((r) => r.plugins);
    if (pluginsToInstall?.length === 0)
        return undefined;
    const updatePackageScripts = await (0, enquirer_1.prompt)([
        {
            name: 'updatePackageScripts',
            type: 'autocomplete',
            message: `Do you want to start using Nx in your package.json scripts?`,
            choices: [
                {
                    name: 'Yes',
                },
                {
                    name: 'No',
                },
            ],
            initial: 0,
        },
    ]).then((r) => r.updatePackageScripts === 'Yes');
    return {
        plugins: pluginsToInstall,
        updatePackageScripts,
    };
}
