// nx/src/config/workspaces.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.toProjectName = exports.Workspaces = void 0;
const path_1 = require("path");
const nx_json_1 = require("./nx-json");
const retrieve_workspace_files_1 = require("../project-graph/utils/retrieve-workspace-files");
// TODO(v18): remove this class
/**
 * @deprecated This will be removed in v18. Use {@link readProjectsConfigurationFromProjectGraph} instead.
 */
class Workspaces {
    constructor(root) {
        this.root = root;
    }
    /**
     * @deprecated Use {@link readProjectsConfigurationFromProjectGraph} instead.
     */
    readWorkspaceConfiguration() {
        const nxJson = (0, nx_json_1.readNxJson)(this.root);
        return {
            version: 2,
            projects: (0, retrieve_workspace_files_1.retrieveProjectConfigurationsSync)(this.root, nxJson)
                .projectNodes,
            ...nxJson,
        };
    }
}
exports.Workspaces = Workspaces;
/**
 * Pulled from toFileName in names from @nx/devkit.
 * Todo: Should refactor, not duplicate.
 */
function toProjectName(fileName) {
    const parts = (0, path_1.dirname)(fileName).split(/[\/\\]/g);
    return parts[parts.length - 1].toLowerCase();
}
exports.toProjectName = toProjectName;
