// nx/src/command-line/run/executor-utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getExecutorInformation = exports.normalizeExecutorSchema = void 0;
const path_1 = require("path");
const nx_plugin_1 = require("../../utils/nx-plugin");
const fileutils_1 = require("../../utils/fileutils");
const schema_utils_1 = require("../../config/schema-utils");
const installation_directory_1 = require("../../utils/installation-directory");
function normalizeExecutorSchema(schema) {
    const version = (schema.version ??= 1);
    return {
        version,
        outputCapture: schema.outputCapture ?? version < 2 ? 'direct-nodejs' : 'pipe',
        properties: !schema.properties || typeof schema.properties !== 'object'
            ? {}
            : schema.properties,
        ...schema,
    };
}
exports.normalizeExecutorSchema = normalizeExecutorSchema;
function cacheKey(nodeModule, executor, root) {
    return `${root}:${nodeModule}:${executor}`;
}
const cachedExecutorInformation = {};
function getExecutorInformation(nodeModule, executor, root) {
    try {
        const key = cacheKey(nodeModule, executor, root);
        if (cachedExecutorInformation[key])
            return cachedExecutorInformation[key];
        const { executorsFilePath, executorConfig, isNgCompat } = readExecutorJson(nodeModule, executor, root);
        const executorsDir = (0, path_1.dirname)(executorsFilePath);
        const schemaPath = (0, schema_utils_1.resolveSchema)(executorConfig.schema, executorsDir);
        const schema = normalizeExecutorSchema((0, fileutils_1.readJsonFile)(schemaPath));
        const implementationFactory = (0, schema_utils_1.getImplementationFactory)(executorConfig.implementation, executorsDir);
        const batchImplementationFactory = executorConfig.batchImplementation
            ? (0, schema_utils_1.getImplementationFactory)(executorConfig.batchImplementation, executorsDir)
            : null;
        const hasherFactory = executorConfig.hasher
            ? (0, schema_utils_1.getImplementationFactory)(executorConfig.hasher, executorsDir)
            : null;
        const res = {
            schema,
            implementationFactory,
            batchImplementationFactory,
            hasherFactory,
            isNgCompat,
            isNxExecutor: !isNgCompat,
        };
        cachedExecutorInformation[key] = res;
        return res;
    }
    catch (e) {
        throw new Error(`Unable to resolve ${nodeModule}:${executor}.\n${e.message}`);
    }
}
exports.getExecutorInformation = getExecutorInformation;
function readExecutorJson(nodeModule, executor, root) {
    const { json: packageJson, path: packageJsonPath } = (0, nx_plugin_1.readPluginPackageJson)(nodeModule, root
        ? [root, __dirname, process.cwd(), ...(0, installation_directory_1.getNxRequirePaths)()]
        : [__dirname, process.cwd(), ...(0, installation_directory_1.getNxRequirePaths)()]);
    const executorsFile = packageJson.executors ?? packageJson.builders;
    if (!executorsFile) {
        throw new Error(`The "${nodeModule}" package does not support Nx executors.`);
    }
    const executorsFilePath = require.resolve((0, path_1.join)((0, path_1.dirname)(packageJsonPath), executorsFile));
    const executorsJson = (0, fileutils_1.readJsonFile)(executorsFilePath);
    const executorConfig = executorsJson.executors?.[executor] || executorsJson.builders?.[executor];
    if (!executorConfig) {
        throw new Error(`Cannot find executor '${executor}' in ${executorsFilePath}.`);
    }
    const isNgCompat = !executorsJson.executors?.[executor];
    return { executorsFilePath, executorConfig, isNgCompat };
}
